File: ios/Classes/DynalinksPlugin.swift
import Flutter
import UIKit
import DynalinksSDK

public class DynalinksPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private var eventSink: FlutterEventSink?
    private var initialLink: DeepLinkResult?
    private var initialLinkConsumed = false

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = DynalinksPlugin()

        // Method channel for one-off calls
        let methodChannel = FlutterMethodChannel(
            name: "com.dynalinks.sdk/dynalinks",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        // Event channel for streaming deep links
        let eventChannel = FlutterEventChannel(
            name: "com.dynalinks.sdk/deep_links",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)

        // Register for app delegate callbacks
        registrar.addApplicationDelegate(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":
            handleConfigure(call, result: result)
        case "checkForDeferredDeepLink":
            handleCheckForDeferredDeepLink(result: result)
        case "handleDeepLink":
            handleDeepLink(call, result: result)
        case "getInitialLink":
            handleGetInitialLink(result: result)
        case "reset":
            handleReset(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleConfigure(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let clientAPIKey = args["clientAPIKey"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing clientAPIKey", details: nil))
            return
        }

        let baseURLString = args["baseURL"] as? String
        let baseURL = baseURLString.flatMap { URL(string: $0) } ?? URL(string: "https://dynalinks.app/api/v1")!
        let logLevelString = args["logLevel"] as? String ?? "error"
        let allowSimulator = args["allowSimulatorOrEmulator"] as? Bool ?? false

        let logLevel: DynalinksLogLevel
        switch logLevelString {
        case "none": logLevel = .none
        case "error": logLevel = .error
        case "warning": logLevel = .warning
        case "info": logLevel = .info
        case "debug": logLevel = .debug
        default: logLevel = .error
        }

        do {
            try Dynalinks.configure(
                clientAPIKey: clientAPIKey,
                baseURL: baseURL,
                logLevel: logLevel,
                allowSimulator: allowSimulator
            )
            result(nil)
        } catch let error as DynalinksError {
            result(flutterError(from: error))
        } catch {
            result(FlutterError(code: "UNKNOWN", message: error.localizedDescription, details: nil))
        }
    }

    private func handleCheckForDeferredDeepLink(result: @escaping FlutterResult) {
        Task {
            do {
                let deepLinkResult = try await Dynalinks.checkForDeferredDeepLink()
                DispatchQueue.main.async {
                    result(self.encodeResult(deepLinkResult, isDeferred: true))
                }
            } catch let error as DynalinksError {
                DispatchQueue.main.async {
                    result(self.flutterError(from: error))
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "UNKNOWN", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func handleDeepLink(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let uriString = args["uri"] as? String,
              let url = URL(string: uriString) else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing or invalid URI", details: nil))
            return
        }

        Task {
            do {
                let deepLinkResult = try await Dynalinks.handleUniversalLink(url: url)
                DispatchQueue.main.async {
                    result(self.encodeResult(deepLinkResult, isDeferred: false))
                }
            } catch let error as DynalinksError {
                DispatchQueue.main.async {
                    result(self.flutterError(from: error))
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "UNKNOWN", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func handleGetInitialLink(result: @escaping FlutterResult) {
        guard !initialLinkConsumed else {
            result(nil)
            return
        }
        initialLinkConsumed = true

        if let link = initialLink {
            result(encodeResult(link, isDeferred: false))
        } else {
            result(nil)
        }
    }

    private func handleReset(result: @escaping FlutterResult) {
        Dynalinks.reset()
        initialLink = nil
        initialLinkConsumed = false
        result(nil)
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    // MARK: - App Delegate

    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([Any]) -> Void
    ) -> Bool {
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
              let url = userActivity.webpageURL else {
            return false
        }

        handleIncomingLink(url)
        return true
    }

    private func handleIncomingLink(_ url: URL) {
        Task {
            do {
                let result = try await Dynalinks.handleUniversalLink(url: url)
                DispatchQueue.main.async {
                    if let sink = self.eventSink {
                        sink(self.encodeResult(result, isDeferred: false))
                    } else {
                        // Store for getInitialLink if no listener yet
                        self.initialLink = result
                    }
                }
            } catch {
                // Log error but don't crash
                print("Dynalinks: Failed to handle Universal Link: \(error)")
            }
        }
    }

    // MARK: - Helpers

    private func encodeResult(_ result: DeepLinkResult, isDeferred: Bool) -> [String: Any?] {
        var dict: [String: Any?] = [
            "matched": result.matched,
            "confidence": result.confidence?.rawValue,
            "match_score": result.matchScore,
            "is_deferred": isDeferred,
        ]

        if let link = result.link {
            dict["link"] = encodeLinkData(link)
        }

        return dict
    }

    private func encodeLinkData(_ link: DeepLinkResult.LinkData) -> [String: Any?] {
        return [
            "id": link.id,
            "name": link.name,
            "path": link.path,
            "shortened_path": link.shortenedPath,
            "url": link.url?.absoluteString,
            "full_url": link.fullURL?.absoluteString,
            "deep_link_value": link.deepLinkValue,
            "ios_deferred_deep_linking_enabled": link.iosDeferredDeepLinkingEnabled,
            "android_fallback_url": link.androidFallbackURL?.absoluteString,
            "ios_fallback_url": link.iosFallbackURL?.absoluteString,
            "enable_forced_redirect": link.enableForcedRedirect,
            "social_title": link.socialTitle,
            "social_description": link.socialDescription,
            "social_image_url": link.socialImageURL?.absoluteString,
            "clicks": link.clicks,
        ]
    }

    private func flutterError(from error: DynalinksError) -> FlutterError {
        let code: String
        let message: String

        switch error {
        case .notConfigured:
            code = "NOT_CONFIGURED"
            message = error.errorDescription ?? "SDK not configured"
        case .invalidAPIKey(let msg):
            code = "INVALID_API_KEY"
            message = msg
        case .simulator:
            code = "SIMULATOR"
            message = error.errorDescription ?? "Not available on simulator"
        case .networkError:
            code = "NETWORK_ERROR"
            message = error.errorDescription ?? "Network error"
        case .invalidResponse:
            code = "INVALID_RESPONSE"
            message = error.errorDescription ?? "Invalid response"
        case .serverError(let statusCode, let msg):
            code = "SERVER_ERROR"
            message = msg ?? "Server error: \(statusCode)"
        case .noMatch:
            code = "NO_MATCH"
            message = error.errorDescription ?? "No match"
        }

        return FlutterError(code: code, message: message, details: nil)
    }
}

File: android/src/main/kotlin/com/dynalinks/dynalinks/DynalinksPlugin.kt
package com.dynalinks.dynalinks

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.dynalinks.sdk.DeepLinkResult
import com.dynalinks.sdk.Dynalinks
import com.dynalinks.sdk.DynalinksError
import com.dynalinks.sdk.DynalinksLogLevel
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.NewIntentListener
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class DynalinksPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
    EventChannel.StreamHandler, NewIntentListener {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var context: Context? = null
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var eventSink: EventChannel.EventSink? = null
    private var initialLink: DeepLinkResult? = null
    private var initialLinkConsumed = false

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext

        methodChannel = MethodChannel(binding.binaryMessenger, "com.dynalinks.sdk/dynalinks")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "com.dynalinks.sdk/deep_links")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        scope.cancel()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "configure" -> handleConfigure(call, result)
            "checkForDeferredDeepLink" -> handleCheckForDeferredDeepLink(result)
            "handleDeepLink" -> handleDeepLink(call, result)
            "getInitialLink" -> handleGetInitialLink(result)
            "reset" -> handleReset(result)
            else -> result.notImplemented()
        }
    }

    // MARK: - Method Handlers

    private fun handleConfigure(call: MethodCall, result: Result) {
        val ctx = context ?: run {
            result.error("NO_CONTEXT", "Application context not available", null)
            return
        }

        val clientAPIKey = call.argument<String>("clientAPIKey") ?: run {
            result.error("INVALID_ARGUMENTS", "Missing clientAPIKey", null)
            return
        }

        val baseURL = call.argument<String>("baseURL") ?: "https://dynalinks.app/api/v1"
        val logLevelString = call.argument<String>("logLevel") ?: "error"
        val allowEmulator = call.argument<Boolean>("allowSimulatorOrEmulator") ?: false

        val logLevel = when (logLevelString) {
            "none" -> DynalinksLogLevel.NONE
            "error" -> DynalinksLogLevel.ERROR
            "warning" -> DynalinksLogLevel.WARNING
            "info" -> DynalinksLogLevel.INFO
            "debug" -> DynalinksLogLevel.DEBUG
            else -> DynalinksLogLevel.ERROR
        }

        try {
            Dynalinks.configure(
                context = ctx,
                clientAPIKey = clientAPIKey,
                baseURL = baseURL,
                logLevel = logLevel,
                allowEmulator = allowEmulator
            )
            result.success(null)
        } catch (e: IllegalArgumentException) {
            result.error("INVALID_API_KEY", e.message, null)
        } catch (e: Exception) {
            result.error("UNKNOWN", e.message, null)
        }
    }

    private fun handleCheckForDeferredDeepLink(result: Result) {
        scope.launch {
            try {
                val deepLinkResult = Dynalinks.checkForDeferredDeepLink()
                result.success(encodeResult(deepLinkResult))
            } catch (e: DynalinksError) {
                result.error(errorCode(e), e.message, null)
            } catch (e: Exception) {
                result.error("UNKNOWN", e.message, null)
            }
        }
    }

    private fun handleDeepLink(call: MethodCall, result: Result) {
        val uriString = call.argument<String>("uri") ?: run {
            result.error("INVALID_ARGUMENTS", "Missing URI", null)
            return
        }

        val uri = Uri.parse(uriString)

        scope.launch {
            try {
                val deepLinkResult = Dynalinks.handleAppLink(uri)
                result.success(encodeResult(deepLinkResult.copy(isDeferred = false)))
            } catch (e: DynalinksError) {
                result.error(errorCode(e), e.message, null)
            } catch (e: Exception) {
                result.error("UNKNOWN", e.message, null)
            }
        }
    }

    private fun handleGetInitialLink(result: Result) {
        if (initialLinkConsumed) {
            result.success(null)
            return
        }
        initialLinkConsumed = true

        val link = initialLink
        if (link != null) {
            result.success(encodeResult(link.copy(isDeferred = false)))
        } else {
            result.success(null)
        }
    }

    private fun handleReset(result: Result) {
        Dynalinks.reset()
        initialLink = null
        initialLinkConsumed = false
        result.success(null)
    }

    // MARK: - ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addOnNewIntentListener(this)

        // Check initial intent
        handleIntent(binding.activity.intent)
    }

    override fun onDetachedFromActivity() {
        activityBinding?.removeOnNewIntentListener(this)
        activity = null
        activityBinding = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addOnNewIntentListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityBinding?.removeOnNewIntentListener(this)
        activity = null
        activityBinding = null
    }

    // MARK: - NewIntentListener

    override fun onNewIntent(intent: Intent): Boolean {
        handleIntent(intent)
        return false
    }

    private fun handleIntent(intent: Intent?) {
        val uri = intent?.data ?: return

        scope.launch {
            try {
                val result = Dynalinks.handleAppLink(uri)
                val sink = eventSink
                if (sink != null) {
                    sink.success(encodeResult(result.copy(isDeferred = false)))
                } else {
                    // Store for getInitialLink if no listener yet
                    initialLink = result
                }
            } catch (e: Exception) {
                // Log but don't crash
                android.util.Log.e("DynalinksPlugin", "Failed to handle App Link", e)
            }
        }
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // MARK: - Helpers

    private fun encodeResult(result: DeepLinkResult): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "matched" to result.matched,
            "confidence" to result.confidence?.name?.lowercase(),
            "match_score" to result.matchScore,
            "is_deferred" to result.isDeferred
        )

        result.link?.let { link ->
            map["link"] = hashMapOf<String, Any?>(
                "id" to link.id,
                "name" to link.name,
                "path" to link.path,
                "shortened_path" to link.shortenedPath,
                "url" to link.url,
                "full_url" to link.fullUrl,
                "deep_link_value" to link.deepLinkValue,
                "android_fallback_url" to link.androidFallbackUrl,
                "ios_fallback_url" to link.iosFallbackUrl,
                "enable_forced_redirect" to link.enableForcedRedirect,
                "social_title" to link.socialTitle,
                "social_description" to link.socialDescription,
                "social_image_url" to link.socialImageUrl,
                "clicks" to link.clicks,
                "ios_deferred_deep_linking_enabled" to link.iosDeferredDeepLinkingEnabled,
                "referrer" to link.referrer,
                "provider_token" to link.providerToken,
                "campaign_token" to link.campaignToken
            )
        }

        return map
    }

    private fun errorCode(error: DynalinksError): String = when (error) {
        is DynalinksError.NotConfigured -> "NOT_CONFIGURED"
        is DynalinksError.InvalidAPIKey -> "INVALID_API_KEY"
        is DynalinksError.Emulator -> "EMULATOR"
        is DynalinksError.InvalidIntent -> "INVALID_INTENT"
        is DynalinksError.NetworkError -> "NETWORK_ERROR"
        is DynalinksError.InvalidResponse -> "INVALID_RESPONSE"
        is DynalinksError.ServerError -> "SERVER_ERROR"
        is DynalinksError.NoMatch -> "NO_MATCH"
        is DynalinksError.InstallReferrerUnavailable -> "INSTALL_REFERRER_UNAVAILABLE"
        is DynalinksError.InstallReferrerTimeout -> "INSTALL_REFERRER_TIMEOUT"
    }
}
